tutorials/examples/mRNA_design/MFE_calculator.py

```diff
@@ -0,0 +1,559 @@
+import math
+
+import numpy as np
+from MFE_utils import (
+    TURNER_FE,
+    base_as_int,
+    int_as_base,
+    pair_int_as_str,
+    str_as_pair_int,
+    tuple_as_pair_int,
+)
+
+_PAIRS = (
+    "AU",
+    "CG",
+    "GC",
+    "GU",
+    "UA",
+    "UG",
+)
+
+
+class RNASolutionBase:
+    """An RNA folding solution and working data
+
+    An instance holds the working data used when solving
+    an RNA folding problem, and afterwards to generate.
+    various expressions of the solution.
+    """
+
+    def __init__(self, rna: str):
+        """Create a holder for the RNA folding solution
+
+        The solution is read from the working arrays allocated here.
+
+        U[i,j] is the minimum energy of [i,j] when j is not paired with i.
+        V[i,j] is the minimum energy of [i,j] closed by (i,j). It is based
+            on V[i+1,j-1] (helix stacking) or U[i+1,j-1] (loop closure)
+            whichever gives the lower result, and is math.inf if (i,j) is
+            inadmissible as a pair.
+        K[i,j] is the nucleotide paired with j in computing U[i,j] or -1
+            if the minimum U[i,j] is obtained by leaving j unpaired.
+        M[i,j] is 1 if V[i,j] is based on helix stacking, is 0 if V[i,j] is
+            based on loop closure, and is -1 if (i,j) is inadmissible. This
+            is not a question of V[i+1,j-1] and U[i+1,j-1] values only.
+        """
+        # Allow mixed case bases and space character initially
+        if junk := rna.strip("ACGUacgu "):
+            raise ValueError(f"RNA bases must be from {{ACGU}} found {junk!r}.")
+        # Remove spaces for computation
+        self.rna = rna = "".join(rna.split())
+        self.N = N = len(rna)
+        # We use integer values to represent the 4 bases
+        self.irna = np.fromiter(
+            map(base_as_int, rna.upper()), dtype=np.int8, count=len(rna)
+        )
+        # Storage for partial results
+        self.V = np.zeros((N, N), dtype=np.float32)
+        self.U = np.zeros((N, N), dtype=np.float32)
+        self.K = np.zeros((N, N), dtype=np.int16)
+        self.M = np.zeros((N, N), dtype=np.int8)
+
+    def getU(self, i, j) -> float:
+        return float(self.U[i, j])
+
+    def getV(self, i, j) -> float:
+        return float(self.V[i, j])
+
+    def getK(self, i, j) -> int:
+        return int(self.K[i, j])
+
+    def getM(self, i, j) -> int:
+        return int(self.M[i, j])
+
+    def energy(self):
+        """The free energy of the minimum energy solution"""
+        n = self.N - 1
+        return min(self.U[0, n], self.V[0, n])
+
+
+class RNASolutionBase2(RNASolutionBase):
```

Comment:
I'm not sure I understand the motivation of these classes being split up into `RNASolutionBase1` and `RNASolutionBase2`.